### PR TITLE
Accessors to own CID within mbedtls_ssl_context

### PR DESCRIPTION
**Pull request by paul-elliott-arm**
## Description
Add accessors to own CID within mbedtls_ssl_context, as per request in #5184

Add sanity tests for all of the own CID set / get/config functions.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
 NO

## Todos
- [x] Tests
- [x] Changelog updated

## Steps to test or reproduce
test_suite_ssl should run clean

### CODE REVIEW

File: library/ssl_tls.c [hunk 1]
```diff
@@ -112,6 +112,40 @@ int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
...
 
+int mbedtls_ssl_get_own_cid( mbedtls_ssl_context *ssl,
+                              int *enabled,
+                              unsigned char own_cid[MBEDTLS_SSL_CID_OUT_LEN_MAX],
+                              size_t *own_cid_len )
+{
+    *enabled = MBEDTLS_SSL_CID_DISABLED;
+
+    if( ssl->conf->transport != MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+    {
```
Comment (mprse) [inline library/ssl_tls.c:123]: Style: braces can be removed.

File: library/ssl_tls.c [hunk 2]
```diff
@@ -112,6 +112,40 @@ int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
...
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    /* We report MBEDTLS_SSL_CID_DISABLED in case the CID length is
+     * zero as this is is indistinguishable from not requesting to
+     * use the CID extension. */
+    if( ssl->own_cid_len == 0 || ssl->negotiate_cid == MBEDTLS_SSL_CID_DISABLED )
+    {
```
Comment (mprse) [inline library/ssl_tls.c:131]: Style: braces can be removed.

File: library/ssl_tls.c [hunk 3]
```diff
@@ -112,6 +112,40 @@ int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
...
+        return( 0 );
+    }
+
+    if( own_cid_len != NULL )
+    {
+        *own_cid_len = ssl->own_cid_len;
+        if( own_cid != NULL )
+        {
```
Comment (mprse) [inline library/ssl_tls.c:139]: Style: braces can be removed.

File: library/ssl_tls.c [hunk 4]
```diff
@@ -112,6 +112,40 @@ int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
...
+{
+    *enabled = MBEDTLS_SSL_CID_DISABLED;
+
+    if( ssl->conf->transport != MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+    {
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    /* We report MBEDTLS_SSL_CID_DISABLED in case the CID length is
+     * zero as this is is indistinguishable from not requesting to
```
Comment (mprse) [inline library/ssl_tls.c:128]: typo: `is is` -> `is`

**Review by mprse — CHANGES_REQUESTED** [review]
Looks good.
Only few style issues and one typo.

**Review by mprse — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
@paul-elliott-arm Unfortunately this has conflicts now that #5622 has been merged (add-add, easy to resolve). Could you rebase to resolve them?

**Comment by mpg** [comment]
Check-files now complaining about missing newline at end of `tests/suites/test_suite_ssl.data` - probably a side-effect of resolving a conflict in that area. (The only other CI issue is an instance of #5583.)

File: ChangeLog.d/add_own_cid_accessors [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Features
+   * Add the function mbedtls_ssl_get_own_cid() to access our own certificate
```
Comment (mpg) [inline ChangeLog.d/add_own_cid_accessors:2]: s/certificate/connection/

File: include/mbedtls/ssl.h [hunk 6]
```diff
@@ -2023,6 +2023,40 @@ int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
...
+ * \brief              Get information about our request for usage of the CID
+ *                     extension in the current connection.
+ *
+ * \param ssl          The SSL context to query.
+ * \param enabled      The address at which to store whether the CID extension
+ *                     is requested to be used or not. If the CID is
+ *                     requested, `*enabled` is set to
+ *                     MBEDTLS_SSL_CID_ENABLED; otherwise, it is set to
+ *                     MBEDTLS_SSL_CID_DISABLED.
+ * \param own_cid     The address of the buffer in which to store our own
```
Comment (mpg) [inline include/mbedtls/ssl.h:2036]: Style: "The" should be aligned with the rest (same for the next param).

File: include/mbedtls/ssl.h [hunk 7]
```diff
@@ -2023,6 +2023,40 @@ int mbedtls_ssl_set_cid( mbedtls_ssl_context *ssl,
                          unsigned char const *own_cid,
                          size_t own_cid_len );
 
+/**
+ * \brief              Get information about our request for usage of the CID
```
Comment (mpg) [inline include/mbedtls/ssl.h:2027]: Ok, now I'm confused about what we're trying to do exactly. Should this function return what we requested, or what we negociated? I'll ask in #5184.

**Review by mpg — COMMENTED** [review]
Partial review, interrupted because I'm no longer sure I understand what it is that we're trying to do here.

File: include/mbedtls/ssl.h (hunk 7 above)
Comment (mpg) [reply include/mbedtls/ssl.h:2027]: Ok, seems returning what we requested is indeed what the original reporter asked for. (Which IMO makes this more an optimisation request than a regression fix, but well, we started it, so let's finish it.)

File: tests/suites/test_suite_ssl.data [hunk 8]
```diff
@@ -3213,3 +3213,6 @@ conf_group:
 
 Test accessor into timing_delay_context
 timing_final_delay_accessor
+
+Sanity test cid functions
+cid_sanity:
```
Comment (mpg) [inline tests/suites/test_suite_ssl.data:3218]: Missing EOL character here.

**Review by mpg — CHANGES_REQUESTED** [review]
Completed my review, no further feedback beyond what I pointed our last time. I like the test, and I see you also fixed some pre-existing test gaps with `mbedtls_set_cid()` while at it, thanks for that!

**Comment by paul-elliott-arm** [comment]
CI Failures are unfortunately the OpenCI results, which has not yet been re-configured. Internal CI passed Ok.

**Review by mpg — APPROVED** [review]
LGTM!

**Review by mprse — APPROVED** [review]

**Pull request merged**